README.md
# MISULTIN

Misultin is an HTTP(S) library which can easily be embedded in your own application. 

https://github.com/ostinelli/misultin

`>-|-|-(°>`


# Features

 * _Very_ fast
 * HTTP and HTTPS
 * Supports multiple **Websocket** protocols (draft-hixie-68, draft-hixie-76, draft-hybi-10 and draft-hybi-17)
 * **Cookies**
 * **Session** Variables
 * Allows for Chunked and Byte **streaming responses**
 * Allows for **streaming file upload** (via Chunked Encoding)
 * Can serves static files from a ```static``` directory (though in production you should consider a specific server such as [nginx](http://nginx.org/) to do so)
 * Has Many customization options (maximum allowed connections, maximum body size, ...)
 * Has a callback function for your **logging** needs
 * Supports **Unicode**
 * Can start multiple servers on a single node
 * Can be used with or without Parametrized Modules
 * Can traps the client closing a browser in Comet applications
 * It's very easy to use


# Quick Start

 The typical 'Hello World" example code is:

```erlang

-module(misultin_hello_world).
-export([start/0, stop/0]).

% start misultin http server
start() ->
    misultin:start_link([{port, 8080}, {loop, fun(Req) -> handle_http(Req) end}]).

% stop misultin
stop() ->
    misultin:stop().

% callback on request received
handle_http(Req) ->
    Req:ok("Hello World.").
```

Issuing the ```start/0``` command will start an HTTP server on port 8080, which will respond to every request with an "Hello World" text.

# Examples

Misultin comes [packed with examples](https://github.com/ostinelli/misultin/tree/master/examples/).

#### Simple Examples

 * [Hello World](https://github.com/ostinelli/misultin/tree/master/examples/misultin_hello_world.erl)
 * [Querystring Variables](https://github.com/ostinelli/misultin/tree/master/examples/misultin_get_variable.erl)
 * [Querystring and POST Variables](https://github.com/ostinelli/misultin/tree/master/examples/misultin_echo.erl)
 * [REST](https://github.com/ostinelli/misultin/tree/master/examples/misultin_rest.erl)
 * [Set and Get Cookies](https://github.com/ostinelli/misultin/tree/master/examples/misultin_cookies_example.erl)
 * [Set and Get Session Variables](https://github.com/ostinelli/misultin/tree/master/examples/misultin_sessions_example.erl)
 * [Serve a Static file for download](https://github.com/ostinelli/misultin/tree/master/examples/misultin_file.erl)
 * [Serving files from a Static directory](https://github.com/ostinelli/misultin/tree/master/examples/misultin_static.erl)
 * [File Upload](https://github.com/ostinelli/misultin/tree/master/examples/misultin_file_upload.erl)
 * [HTTPS example](https://github.com/ostinelli/misultin/tree/master/examples/misultin_ssl.erl)
 * [Performing a simple redirection](https://github.com/ostinelli/misultin/tree/master/examples/misultin_redirect.erl)
 * [Serving compressed content](https://github.com/ostinelli/misultin/tree/master/examples/misultin_compress.erl)
 * [Logging Access](https://github.com/ostinelli/misultin/tree/master/examples/misultin_access_log.erl)

#### Websockets

 * [Simple Websocket](https://github.com/ostinelli/misultin/tree/master/examples/misultin_websocket_example.erl)
 * [Simple Websocket on SSL](https://github.com/ostinelli/misultin/tree/master/examples/misultin_websocket_example_ssl.erl)
 * [Websocket exposing the close event](https://github.com/ostinelli/misultin/tree/master/examples/misultin_websocket_event_example.erl)
 * [Websocket exposing the close event, example 2](https://github.com/ostinelli/misultin/tree/master/examples/misultin_websocket_event_example2.erl)
 * [Access Session Variables from Websockets](https://github.com/ostinelli/misultin/tree/master/examples/misultin_websocket_sessions_example.erl)


#### Comets

 * [Long Polling](https://github.com/ostinelli/misultin/tree/master/examples/misultin_comet_long_polling.erl)
 * [iFrame Technique](https://github.com/ostinelli/misultin/tree/master/examples/misultin_comet_iframe.erl)
 * [iFrame Technique using the close event](https://github.com/ostinelli/misultin/tree/master/examples/misultin_comet_iframe_event.erl)


#### More Advanced

 * [Sending Chunked Content](https://github.com/ostinelli/misultin/tree/master/examples/misultin_chunked.erl)
 * [Sending Byte Streaming Content](https://github.com/ostinelli/misultin/tree/master/examples/misultin_stream.erl)
 * [Receiving endless streaming Upload](https://github.com/ostinelli/misultin/tree/master/examples/misultin_body_recv.erl)
 * [Unicode](https://github.com/ostinelli/misultin/tree/master/examples/misultin_unicode.erl)
 * [REST with UTF-8](https://github.com/ostinelli/misultin/tree/master/examples/misultin_rest_utf8.erl)
 * [Starting a nameless server](https://github.com/ostinelli/misultin/tree/master/examples/misultin_hello_world_nameless.erl)
 * [Starting multiple servers on a same node, with a custom name](https://github.com/ostinelli/misultin/tree/master/examples/misultin_multiple_servers_custom_name.erl)
 * [Using the HAProxy protocol](https://github.com/ostinelli/misultin/tree/master/examples/misultin_proxy_protocol.erl)

# Module Exports

The complete list of module exports can be found [here](https://github.com/ostinelli/misultin/tree/master/EXPORTS.md).

# Parametrized modules

Some developers hate them, some love them. Misultin allows you to choose if you want to use them or not. The same **Hello World** example shown here above, but without parametrized modules, looks like this:

```erlang

-module(misultin_hello_world).
-export([start/0, stop/0]).

% start misultin http server
start() ->
    misultin:start_link([{port, 8080}, {loop, fun(Req) -> handle_http(Req) end}]).

% stop misultin
stop() ->
    misultin:stop().

% callback on request received
handle_http(Req) ->
    misultin_req:ok("Hello World.", Req).
```

# Dependencies

You will need:

 * [Erlang](http://www.erlang.org/download.html) >= R14B01
 * [Rebar](https://github.com/basho/rebar) to compile

# Under the hood
Misultin is built using the OTP principles. When you start it using the ```misultin:start_link/1``` command, you are actually starting a supervisor which handles all of Misultin's servers and modules.

Therefore, in real life applications you should always embed it in your own application. An easy example on how this can be done can be found in the Application Example [here](https://github.com/ostinelli/misultin/tree/master/examples/application).

# SSL Notes

If you are running misultin behind an SSL terminator such as stunnel or stud, and are using websockets, to make the websocket handshakes work, set in the starting options:

```erlang
{ws_force_ssl, true} 
```

If you are using stunnel to terminate, to make misultin expect a PROXY.. line as per the [proxy protocol](http://haproxy.1wt.eu/download/1.5/doc/proxy-protocol.txt) you can also set in the starting options:
```erlang
{proxy_protocol, true}
```

Newer versions of stunnel support this with the "protocol = proxy" config option.

CHANGELOG.md
# CHANGELOG

### 0.9:
 * added SESSIONS state, persistent across requests
 * added access log callback function, so that main application can log HTTP access
 * added streaming input for big files or endless input, using a manual ```body_recv function``` in conjunction with the ```{auto_recv_body, false}``` option
 * added static directory support, so that GET requests to /static/* can automatically send files from a specified directory (thanks to egobrain suggestion)
 * added request redirection helper method
 * consistently improved memory usage by not copying by default to handler processes the full request or websocket record
 * added configuration option to set which websocket versions must be supported by the server
 * added support for websocket draft-hybi-10
 * added support for websocket draft-hybi-17 (thanks to RJ)
 * added support for websockets on FireFox (thanks to Egobrain)
 * added support for 'If-Modified-Since' headers in file sending (thanks to davidgaleano)
 * added support for websockets when behind stunnel with ```{external_ssl, boolean()}``` option (thanks to RJ)
 * added support to see the correct client IP when behind stunnel, according to http://haproxy.1wt.eu/download/1.5/doc/proxy-protocol.txt (thanks to RJ)
 * added support for OPTIONS method (thanks to majek)
 * rebar-ized makefile
 * corrected minor bugs (thank you all - you know who you are!)

### 0.8:
 * Misultin has been redesigned to use supervisor behaviours where appropriate, to be more OTP compliant
 * added Cookie support
 * added preliminary support multipart/form-data and a file upload example (thanks to Max Lapshin)
 * added Date and Server headers
 * added support for headers being specified as binaries (thanks to akaspin)
 * added an example on how to properly embed misultin in your application
 * ```Req:get(peer_addr)``` now properly extracts peer information from headers "X-Real-Ip" or "X-Forwarded-For" if these are available (thanks to Max Lapshin)
 * solved bug on large data being sent over via websockets (thanks to omarkj)
 * corrected binary sending bug in websockets which would fail binaries on io_lib format (thanks to normanb)
 * added recbuf advanced option (issue #40)
 * added preliminary test suite
 * various optimizations using binary when needed

### 0.7.1:
 * considerably improved stability under heavy load
 * misultin now accepts incoming connections with a pool of acceptors instead of a single one
 * Misultin can now be used both with parametrized modules and with pure erlang code too (thanks to yrashk, nox and essen)
 * added support for HEAD, PUT, DELETE, TRACE and CONNECT methods
 * now websockets are on ```{active, once}``` mode to avoid malicious clients overflooding (thanks to essen)
 * ensured that body of request can be read on all methods except TRACE as per http specs
 * added support for iolist() in chunked resposes (thanks to RJ)

### 0.7:
 * added ```max_connections options``` parameter, which specifies maximum concurrent open connections accepted by the server
 * added ```post_max_size``` options parameter, which sets the maximum size of POST data
 * added ```get_url_max_size``` options parameter, which sets the maximum length of URI
 * added CHUNKED support, both for incoming requests and outgoing responses (thanks to yrashk suggestion)
 * added trapping of client closing a browser in Comet applications (thanks to yrashk)
 * added SSL support for websockets (enhancement track #25, thanks to viplifes)
 * Misultin can now be started without a registered name or with a different name, so that multiple versions of misultin can be started on a single node
 * added support for IP address specified in tuple format (thanks to okeuday suggestion)
 * added support to extract plain uri unquoted as a list() (thanks to okeuday)
 * added Comet Long Polling example
 * added Comet iFrame example
 * added the killing of alive processes on server shutdown
 * the GET uri parameters are now also available on POST requests
 * added custom headers on file sending to browser (thanks to josevalim)
 * additional minor adjustments

### 0.6.2:
 * refactored to considerably improve sending of static files
 * minor bug corrections

### 0.6.1:
 * added support to websocket protocol hixie draft 76 (thanks to sergio veiga)
 * added support to multiple websocket draft protocols (for backwards compatibility)
 * added ```ws_autoexit option``` which allows to get an event on websocket controlling processes (issue track #15, suggestion of esente)
 * added headers also in misultin websockets (thanks to jlirochon)
 * made it basho's rebar friendly (thanks to mrinalwadhwa)

### 0.6:
 * added HTTP compression option
 * refactoring of the main server loop, so that it is now isolated from the HTTP functionality
 * removed unnecessary compilation warnings
 * replaced ```proplists:get_value``` with much faster utility function

### 0.5:
 * added SSL support
 * refactoring of the acceptor loop

### 0.4:
 * added preliminary websocket support

### 0.3.4:
 * added Req support to return the socket handling the request
 * bug correction on Content-Length: 0 header causing timeout on POST requests (issue track #12, thanks to gdamjan)

### 0.3.3:
 * added echoing of the Connection header (issue track #7, thanks to thijsterlouw)
 * bug correction on acceptor respawning (issue track #10, thanks to thijsterlouw)

### 0.3.2:
 * optimized error handling (issue track #5, thanks to Max Lapshin)

### 0.3.1:
 * added flow control using inet options {active, once} (issue track #4, thanks to Max Lapshin)
 * added support to standard http headers response
 * added http 400 bad request error in socket handling
 * bug correction: removed erroneous sending of response timeout on listening open connections
 * added stream_support optimization option

### 0.3:
 * reengineering of the listener process, using active instead of passive mode in request parsing, except for BODY where passive is still used (thanks to Lev Walkin)

src/misultin.erl
% ==========================================================================================================
% MISULTIN - Main Supervisor
%
% >-|-|-(°>
% 
% Copyright (C) 2011, Roberto Ostinelli <roberto@ostinelli.net>.
% All rights reserved.
%
% Code portions from Sean Hinde have been originally taken under BSD license from Trapexit at the address:
% <http://www.trapexit.org/A_fast_web_server_demonstrating_some_undocumented_Erlang_features>
%
% BSD License
% 
% Redistribution and use in source and binary forms, with or without modification, are permitted provided
% that the following conditions are met:
%
%  * Redistributions of source code must retain the above copyright notice, this list of conditions and the
%	 following disclaimer.
%  * Redistributions in binary form must reproduce the above copyright notice, this list of conditions and
%	 the following disclaimer in the documentation and/or other materials provided with the distribution.
%  * Neither the name of the authors nor the names of its contributors may be used to endorse or promote
%	 products derived from this software without specific prior written permission.
%
% THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED
% WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
% PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR
% ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED
% TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
% HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
% NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
% POSSIBILITY OF SUCH DAMAGE.
% ==========================================================================================================
-module(misultin).
-behaviour(supervisor).
-vsn("0.9").

% API
-export([start_link/1, stop/0, stop/1]).

% supervisor callbacks
-export([init/1]).

% macros
-define(SERVER, ?MODULE).

% includes
-include("../include/misultin.hrl").


% ============================ \/ API ======================================================================

% Starts the server.
-spec start_link(Options::gen_proplist()) -> {ok, Pid::pid()} | {error, Reason::term()}.
start_link(Options) when is_list(Options) ->
	?LOG_DEBUG("starting misultin server",[]),
	% check if name option has been specified, otherwise default to 'misultin' as regname
	case get_option({name, ?SERVER, fun is_atom/1, invalid_misultin_process_name}, Options) of
		{error, Reason} ->
			% option error
			?LOG_ERROR("error in name option: ~p", [Reason]),
			{error, Reason};
		{name, false} ->
			% start misultin without name
			?LOG_DEBUG("starting misultin without a registered name",[]),
			supervisor:start_link(?MODULE, [Options]);
		{name, Value} ->
			% start misultin with specified name
			?LOG_DEBUG("starting misultin with registered name ~p", [Value]),
			Options0 = lists:keyreplace(name, 1, Options, {name, Value}),
			supervisor:start_link({local, Value}, ?MODULE, [Options0])
	end.

% Manually stops the server.
-spec stop() -> true.
-spec stop(SupRef::atom() | pid()) -> true.
stop() ->
	stop(?SERVER).
stop(undefined) ->
	true;
stop(SupName) when is_atom(SupName) ->
	stop(whereis(SupName));
stop(SupPid) when is_pid(SupPid) ->
	?LOG_INFO("shutting down misultin server",[]),
	exit(SupPid, normal).

% ============================ /\ API ======================================================================


% ============================ \/ SUPERVISOR CALLBACKS =====================================================

% ----------------------------------------------------------------------------------------------------------
% Function: -> {ok,  {SupFlags,  [ChildSpec]}} | ignore | {error, Reason}
% Description: Starts the supervisor
% ----------------------------------------------------------------------------------------------------------å
-spec init(Options::gen_proplist()) -> {ok, term()} | {error, Reason::term()}.
init([Options]) ->
	?LOG_DEBUG("starting server with pid: ~p", [self()]),
	% test and get options
	OptionProps = [
		% socket
		{ip, {0, 0, 0, 0}, fun check_and_convert_string_to_ip/1, invalid_ip},
		{port, 80, fun is_non_neg_integer/1, port_not_integer},
		{backlog, 128, fun is_non_neg_integer/1, backlog_not_integer},
		{acceptors_poolsize, 10, fun is_non_neg_integer/1, invalid_acceptors_poolsize_option},
		{recv_timeout, 30*1000, fun is_non_neg_integer/1, recv_timeout_not_integer},
		{max_connections, 4096, fun is_non_neg_integer/1, invalid_max_connections_option},
		{ssl, false, fun check_ssl_options/1, invalid_ssl_options},
		{ws_force_ssl, false, fun is_boolean/1, invalid_ssl_external},
		{proxy_protocol, false, fun is_boolean/1, invalid_proxy_protocol},
		{recbuf, default, fun check_recbuf/1, recbuf_not_integer},
		% misultin
		{post_max_size, 4*1024*1024, fun is_non_neg_integer/1, invalid_post_max_size_option},		% defaults to 4 MB
		{get_url_max_size, 2000, fun is_non_neg_integer/1, invalid_get_url_max_size_option},
		{compress, false, fun is_boolean/1, invalid_compress_option},
		{loop, {error, undefined_loop}, fun is_function/1, loop_not_function},
		{autoexit, true, fun is_boolean/1, invalid_autoexit_option},
		{ws_loop, undefined, fun is_function/1, ws_loop_not_function},
		{ws_autoexit, true, fun is_boolean/1, invalid_ws_autoexit_option},
		{ws_versions, ['draft-hybi-17', 'draft-hybi-10', 'draft-hixie-76'], fun check_ws_version/1, unsupported_ws_vsn_specified},
		{sessions_expire, 600, fun is_non_neg_integer/1, invalid_sessions_expire_option},
		{access_log, undefined, fun check_access_log/1, invalid_access_log_option},
		{auto_recv_body, true, fun is_boolean/1, invalid_auto_recv_body_option},
		{static, false, fun check_static/1, invalid_static_option}		% if set to a directory, requests to /static/* will automatically send files from the directory
	],
	OptionsVerified = lists:foldl(fun(OptionProp, Acc) -> [get_option(OptionProp, Options)|Acc] end, [], OptionProps),
	case proplists:get_value(error, OptionsVerified) of
		undefined ->
			% ok, no error found in options
			% tcp options
			Ip = proplists:get_value(ip, OptionsVerified),
			Port = proplists:get_value(port, OptionsVerified),
			Backlog = proplists:get_value(backlog, OptionsVerified),
			AcceptorsPoolsize = proplists:get_value(acceptors_poolsize, OptionsVerified),
			RecvTimeout = proplists:get_value(recv_timeout, OptionsVerified),
			MaxConnections = proplists:get_value(max_connections, OptionsVerified),
			SslOptions0 = proplists:get_value(ssl, OptionsVerified),
			WsForceSsl = proplists:get_value(ws_force_ssl, OptionsVerified),
			ProxyProtocol = proplists:get_value(proxy_protocol, OptionsVerified),
			% misultin options
			PostMaxSize = proplists:get_value(post_max_size, OptionsVerified),
			GetUrlMaxSize = proplists:get_value(get_url_max_size, OptionsVerified),
			Compress = proplists:get_value(compress, OptionsVerified),
			Loop = proplists:get_value(loop, OptionsVerified),
			AutoExit = proplists:get_value(autoexit, OptionsVerified),
			WsLoop = proplists:get_value(ws_loop, OptionsVerified),
			WsAutoExit = proplists:get_value(ws_autoexit, OptionsVerified),
			WsVersions = proplists:get_value(ws_versions, OptionsVerified),
			RecBuf = proplists:get_value(recbuf, OptionsVerified),
			SessionsExpireSec = proplists:get_value(sessions_expire, OptionsVerified),
			AccessLogFun = proplists:get_value(access_log, OptionsVerified),
			AutoRecvBody = proplists:get_value(auto_recv_body, OptionsVerified),
			Static = proplists:get_value(static, OptionsVerified),
			% set additional options according to socket mode if necessary
			Continue = case SslOptions0 of
				false ->
					% without SSL
					SocketMode = http,
					InetOpt = case Ip of
						{_, _, _, _} ->
							% IPv4
							inet;
						{_, _, _, _, _, _, _, _} ->
							% IPv6
							inet6
					end,
					AdditionalOptions = [InetOpt],
					true;
				_ ->
					% with SSL
					SocketMode = ssl,
					% the only current way to use {active, once} in Ssl is to start the crypto module
					% and set {ssl_imp, new} as SSL option, see
					% <http://www.erlang.org/cgi-bin/ezmlm-cgi?4:mss:50633:201004:fpopocbfkpppecdembbe>
					AdditionalOptions = [{ssl_imp, new}|SslOptions0],
					% start Ssl and crypto applications if necessary, and get outcomes
					AppStartResults = lists:keyfind(error, 1, [start_application(crypto), start_application(public_key), start_application(ssl)]),
					case AppStartResults of
						false ->
							% all applications started succesfully
							true;
						_ ->
							% error starting application
							{error, AppStartResults}
					end
			end,
			% proceed?
			case Continue of
				true ->
					% set options
					OptionsTcp0 = [binary, {packet, raw}, {ip, Ip}, {reuseaddr, true}, {active, false}, {backlog, Backlog}|AdditionalOptions],
					OptionsTcp = case RecBuf of
						default -> OptionsTcp0;
						_ -> [{recbuf, RecBuf}|OptionsTcp0]
					end,							
					% build custom_opts
					CustomOpts = #custom_opts{
						post_max_size = PostMaxSize,
						get_url_max_size = GetUrlMaxSize,
						compress = Compress,
						loop = Loop,
						autoexit = AutoExit,
						ws_loop = WsLoop,
						ws_autoexit = WsAutoExit,
						ws_versions = WsVersions,
						access_log = AccessLogFun,
						ws_force_ssl = WsForceSsl,
						proxy_protocol = ProxyProtocol,
						auto_recv_body = AutoRecvBody,
						static = Static
					},
					% define misultin_server supervisor specs
					ServerSpec = {server, {misultin_server, start_link, [{MaxConnections}]}, permanent, 60000, worker, [misultin_server]},
					% define sessions supervisor
					SessionsSpec = {sessions, {misultin_sessions, start_link, [{self(), SessionsExpireSec}]}, permanent, 60000, worker, [misultin_sessions]},
					% define acceptors supervisor specs
					AcceptorSupSpec = {acceptors_sup, {misultin_acceptors_sup, start_link, [self(), Port, OptionsTcp, AcceptorsPoolsize, RecvTimeout, SocketMode, CustomOpts]}, permanent, infinity, supervisor, [misultin_acceptors_sup]},
					% ip address
					?LOG_INFO("starting misultin server on address ~s and port ~p", [misultin_utility:convert_ip_to_list(Ip), Port]),
					% spawn
					{ok, {{one_for_all, 5, 30}, [ServerSpec, SessionsSpec, AcceptorSupSpec]}};
				Error ->
					?LOG_ERROR("error starting misultin server: ~p", [Error]),
					{error, Error}
			end;
		Reason ->
			% error found in options
			?LOG_ERROR("option error starting misultin server: ~p", [Reason]),
			{error, Reason}
	end.

% ============================ /\ SUPERVISOR CALLBACKS =====================================================


% ============================ \/ INTERNAL FUNCTIONS =======================================================

% Checks and if necessary converts a string Ip to inet repr.
-spec check_and_convert_string_to_ip(Ip::string() | tuple()) -> inet:ip_address() | false.
check_and_convert_string_to_ip(Ip) when is_tuple(Ip) ->
	case size(Ip) of
		4 ->
			% check for valid ipv4
			LIp = [Num || Num <- tuple_to_list(Ip), Num >= 0, Num =< 255],
			length(LIp) =:= 4 andalso Ip;
		8 ->
			% check for valid ipv6
			LIp = [Num || Num <- tuple_to_list(Ip), Num >= 0, Num =< 16#FFFF],
			length(LIp) =:= 8 andalso Ip;
		_ ->
			false
	end;
check_and_convert_string_to_ip(Ip) ->
	case inet_parse:address(Ip) of
		{error, _Reason} ->
			false;
		{ok, IpTuple} ->
			IpTuple
	end.
	
% Checks if all necessary Ssl Options have been specified
-spec check_ssl_options(SslOptions::gen_proplist()) -> boolean().
check_ssl_options(SslOptions) ->
	Opts = [verify, fail_if_no_peer_cert, verify_fun, depth, certfile, keyfile, password, cacertfile, ciphers, reuse_sessions, reuse_session],
	F = fun({Name, _Value}) ->
		case lists:member(Name, Opts) of
			false -> false;
			_ -> true
		end
	end,
	lists:all(F, SslOptions).

% check if recbuf has been set
-spec check_recbuf(RecBuf::default | non_neg_integer()) -> boolean().
check_recbuf(default) -> default;
check_recbuf(RecBuf) when is_integer(RecBuf), RecBuf > 0 -> true;
check_recbuf(_RecBuf) -> false.

% check if access log fun has been set
-spec check_access_log(AccessLogFun::false | function()) -> boolean().
check_access_log(undefined) -> true;
check_access_log(AccessLogFun) when is_function(AccessLogFun) -> true;
check_access_log(_AccessLogFun) -> false.

% check if ws specified versions are implemented. order does matter so we build a list in proper order
-spec check_ws_version([websocket_version()]) -> false | [websocket_version()].
check_ws_version(WsVsn) ->
	ImplementedVsn = ['draft-hybi-17', 'draft-hybi-10', 'draft-hixie-76', 'draft-hixie-68'],
	%  build an ordered list of supported versions chosen by user.
	F = fun(SupportedVsn, Acc) ->
		case lists:member(SupportedVsn, WsVsn) of
			true -> [SupportedVsn|Acc];
			false -> Acc
		end
	end,
	OrderedVsn = lists:reverse(lists:foldl(F, [], ImplementedVsn)),
	% if length do not agree, then user has specified an unsupported version
	case length(OrderedVsn) =:= length(WsVsn) of
		true -> OrderedVsn;
		_ -> false
	end.

% check if a number is a non negative integer
-spec is_non_neg_integer(term()) -> boolean().
is_non_neg_integer(N) when is_integer(N), N >= 0 -> true;
is_non_neg_integer(_) -> false.

% check if the static option is valid
-spec check_static(list() | false) -> boolean().
check_static(false) -> true;
check_static(Path) when is_list(Path) -> true;
check_static(_) -> false.

% Validate and get misultin options.
-spec get_option({
	OptionName::atom(),
	DefaultValue::term(),
	CheckAndConvertFun::function(),
	FailTypeError::term()
	}, Options::gen_proplist()) -> misultin_option() | {error, Reason::term()}.
get_option({OptionName, DefaultValue, CheckAndConvertFun, FailTypeError}, Options) ->
	case proplists:get_value(OptionName, Options) of
		undefined ->
			case DefaultValue of
				{error, Reason} ->
					{error, Reason};
				Value -> 
					{OptionName, Value}
			end;
		Value ->
			case CheckAndConvertFun(Value) of
				false ->
					{error, {FailTypeError, Value}};
				true -> 
					{OptionName, Value};
				OutValue ->
					{OptionName, OutValue}
			end
	end.

% Start an application.
-spec start_application(Application::atom()) -> ok | {error, Reason::term()}.
start_application(Application) ->
	case lists:keyfind(Application, 1, application:which_applications()) of
		false ->
			?LOG_DEBUG("starting application ~p", [Application]),
			case application:start(Application) of
				ok ->
					ok;
				{error, Reason} ->
					?LOG_ERROR("error starting application ~p", [Application]),
					{error, Reason}
			end;
		_ ->
			?LOG_DEBUG("application ~p is already running, skip", [Application]),
			ok
	end.

% ============================ /\ INTERNAL FUNCTIONS =======================================================
